### PR TITLE
[DSM] PEPPER-676 Fix for Final Scan Error

### PR DESCRIPTION
**Pull request by pegahtah**
This PR is to bring hotfix changes from hotfix branch into develop.
PR for hotfix is [here](https://github.com/broadinstitute/ddp-study-server/pull/2489)

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/ScanError.java [hunk 1]
```diff
@@ -1,14 +1,27 @@
...
+
+    public ScanError(String kit, String error, String shortId) {
+        this.kit = kit;
+        this.error = error;
+        this.shortId = shortId;
+    }
+
+    public boolean isScanErrorOnlyBspParticipantId(String bspCollaboratorParticipantId) {
+        return (StringUtils.isBlank(getError()) && StringUtils.isBlank(getKit()))
+                || (StringUtils.isBlank(getError()) && getShortId().equals(bspCollaboratorParticipantId));
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/ScanError.java]: Are shortIds case sensitive?  If not, maybe equalsIgnoreCase?

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/kits/KitScanErrorTest.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package org.broadinstitute.dsm.kits;
+
+import org.broadinstitute.dsm.model.kit.ScanError;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class KitScanErrorTest {
+
+    @Test
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/kits/KitScanErrorTest.java:9]: Can you please add two additional tests?
One test where for error = null and shortId =null and one for error = null and kit = null;

File: pepper-apis/dsm-core/src/test/integration/org/broadinstitute/dsm/model/kit/KitFinalScanUseCaseIntegrationTest.java [hunk 3]
```diff
@@ -76,4 +78,38 @@ public void processIsBloodKitHasNoTracking() {
...
+        setAndSaveKitRequestId(kitRequestShipping);
+        Integer kitId = kitDao.insertKit(kitRequestShipping);
+
+        kitIds.add(kitId);
+
+        List<ScanPayload> scanPayloads = Arrays.asList(
+                new SentAndFinalScanPayload("ddpLabel", "kitLabelMoreThan14")
+        );
+        DDPInstanceDto ddpInstanceDto = new DDPInstanceDao().getDDPInstanceByInstanceName(TestInstanceCreator.TEST_INSTANCE).orElseThrow();
+        KitPayload kitPayload = new KitPayload(scanPayloads, 94, ddpInstanceDto);
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/test/integration/org/broadinstitute/dsm/model/kit/KitFinalScanUseCaseIntegrationTest.java]: What's the magic number (94) for?

**Comment by otchet-broad** [comment]
Well done @pegahtah!  Thank you for getting this to production so quickly and thoughtfully.

I let just a few comments for some edge cases I see that you might want to account for.

File: pepper-apis/dsm-core/src/test/integration/org/broadinstitute/dsm/model/kit/KitFinalScanUseCaseIntegrationTest.java (hunk 3 above)
Comment (pegahtah) [reply pepper-apis/dsm-core/src/test/integration/org/broadinstitute/dsm/model/kit/KitFinalScanUseCaseIntegrationTest.java]: That's a user id who has all the access. Used in most of kit tests

Comment (otchet-broad) [reply pepper-apis/dsm-core/src/test/integration/org/broadinstitute/dsm/model/kit/KitFinalScanUseCaseIntegrationTest.java]: Mind making that a constant with a descriptive name?

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/kits/KitScanErrorTest.java (hunk 2 above)
Comment (pegahtah) [reply pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/kits/KitScanErrorTest.java:9]: One of those tests actually brought up a weak point and I added an extra check in the method for `shortId = null` case, thank you!!

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/ScanError.java [hunk 4]
```diff
@@ -1,14 +1,27 @@
...
+
+    public ScanError(String kit, String error, String shortId) {
+        this.kit = kit;
+        this.error = error;
+        this.shortId = shortId;
+    }
+
+    public boolean isScanErrorOnlyBspParticipantId(String bspCollaboratorParticipantId) {
+        return (StringUtils.isBlank(getError()) && StringUtils.isBlank(getKit())) || (StringUtils.isBlank(getError())
+                && StringUtils.isNotBlank(getShortId()) && getShortId().equalsIgnoreCase(bspCollaboratorParticipantId));
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/ScanError.java:25]: Huzzah!

File: pepper-apis/dsm-core/src/test/integration/org/broadinstitute/dsm/model/kit/KitBaseUseCaseTest.java [hunk 5]
```diff
@@ -21,6 +21,8 @@ public class KitBaseUseCaseTest {
     protected static final List<Integer> kitTypeIds = new ArrayList<>();
     protected static final TestInstanceCreator testInstance = new TestInstanceCreator();
 
+    protected static int USER_ID = 94;
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/test/integration/org/broadinstitute/dsm/model/kit/KitBaseUseCaseTest.java:24]: Thank you!

**Review by otchet-broad — APPROVED** [review]
Thanks for the updates and test coverage!
